pkg/chunked/storage_linux.go

```diff
@@ -1734,3 +1754,63 @@ func validateChunkChecksum(chunk *internal.FileMetadata, root, path string, offs
 
 	return digester.Digest() == digest
 }
+
+func newStagedFileGetter(dirFD int, differOpts *graphdriver.DifferOptions, entriesWithOriginalNames []fileMetadata) (*stagedFileGetter, error) {
+	var pathMap map[string]string = nil
+	if differOpts != nil && differOpts.Format == graphdriver.DifferOutputFormatFlat {
+		pathMap = make(map[string]string, len(entriesWithOriginalNames))
+		// This needs to match cloneEntriesWithFlatNames
+		for _, e := range entriesWithOriginalNames {
+			if e.Type != TypeReg {
+				continue
+			}
+			if e.Digest == "" {
+				return nil, fmt.Errorf("missing digest for %q", e.Name)
+			}
+			digest, err := digest.Parse(e.Digest)
+			if err != nil {
+				return nil, err
+			}
+			d := digest.Encoded()
+			pathMap[filepath.Clean(e.Name)] = fmt.Sprintf("%s/%s", d[0:2], d[2:])
+		}
+	}
+
+	dirFD2, err := unix.Dup(dirFD)
+	if err != nil {
+		return nil, err
+	}
+	rootDir := os.NewFile(uintptr(dirFD2), "staging root directory")
```

Comment:
This is inelegant, maybe the `*os.File` creation can happen in the caller, making _that_ the primary holder, and the FD integer a derived value? At least the `Dup` should be avoidable.

Reply:
Introduced `dirFile`.

(Eventually, it seems to me that it might be clean to have a `stagingDestination struct { dirFile , …}` holder, with various functions like `openDestinationFile` becoming methods on this new type, instead of passing a raw file descriptor as a parameter all over the place. That would clearly be a separate PR.)